doop.c

```diff
@@ -696,30 +748,27 @@ Perl_do_join(pTHX_ SV *sv, SV *delim, SV **mark, SV **sp)
 
     if (items-- > 0) {
         if (*mark)
-            sv_catsv(sv, *mark);
+            sv = do_join_inner(sv, *mark);
         mark++;
     }
 
     if (delimlen) {
-        const U32 delimflag = DO_UTF8(delim) ? SV_CATUTF8 : SV_CATBYTES;
         for (; items > 0; items--,mark++) {
-            STRLEN len;
-            const char *s;
-            sv_catpvn_flags(sv,delims,delimlen,delimflag);
-            s = SvPV_const(*mark,len);
-            sv_catpvn_flags(sv,s,len,
-                            DO_UTF8(*mark) ? SV_CATUTF8 : SV_CATBYTES);
+            sv = do_join_inner(sv, delim);
+            sv = do_join_inner(sv, *mark);
         }
     }
     else {
-        for (; items > 0; items--,mark++)
-        {
-            STRLEN len;
-            const char *s = SvPV_const(*mark,len);
-            sv_catpvn_flags(sv,s,len,
-                            DO_UTF8(*mark) ? SV_CATUTF8 : SV_CATBYTES);
+        for (; items > 0; items--,mark++) {
+            sv = do_join_inner(sv, *mark);
```

Comment:
and here we call do_join_inner() with the same SV as the lhs. So we can avoid the need to check it for magic each time we call by passing that state in as a flag.

Reply:
Isn't the reason we assign the sv each time is that the return value may not be the same sv, especially in the case where the first element didn't have overload concat magic, but the second element does, after which the lhs gets the concat magic? Checking once would break that, right?

Reply:
No, magic isnt viral like that. See comment below.